### ISSUE TITLE
tfa.optimizers.AveragedOptimizerWrapper(optimizer) does not save the optimizer state when used with tf.train.Checkpoint

### ISSUE THREAD

**Issue by gowithdaflo** [tensorflow/addons#2354]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab (Linux)
- TensorFlow version and how it was installed (source or binary): 2.4.0 (preinstalled)
- TensorFlow-Addons version and how it was installed (source or binary): 0.8.3
- Python version: 3.8
- Is GPU used? (yes/no): yes

**Describe the bug**

I created a moving average optimizer by 
`optimizer = tfa.optimizers.MovingAverage(tf.keras.optimizers.Adam(0.001), average_decay=0.99)`
and later use checkpointing by
`tf.train.Checkpoint(step=tf.Variable(1), optimizer=optimizer, model=model)`

However, when doing it this way the optimizer state of the Adam optimizer state is NOT saved but only the moving average of the variables. Only when explicitly adding the adam optimizer to the checkpoint
`ckpt = tf.train.Checkpoint(step=tf.Variable(1), optimizer=optimizer, _optimizer = optimizer._optimizer,  model=model)`
results in the desired behaviour of also saving the state of the Adam Optimizer.
One can see the difference when analyzing the diffference in the saved ckpt-files.

In my opinion it would be expected that adding the MovingAverage optimizer to the checkpoint also tracks the Adam optimizer.

**Code to reproduce the issue**

`
import numpy as np
import tensorflow as tf
import tensorflow_addons as tfa
from tensorflow.keras import layers

(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

train_labels = train_labels[:1000]
test_labels = test_labels[:1000]

train_images = tf.cast(train_images[:1000][...,None], tf.float32) # (None,28,28,1)
test_images = tf.cast(test_images[:1000][...,None], tf.float32)

class Model(tf.keras.Model):
  def __init__(self):
    super().__init__()
    model = tf.keras.models.Sequential()
    model.add(layers.Conv2D(24, 3, activation='relu'))
    model.add(layers.MaxPooling2D(2))
    model.add(layers.Conv2D(20,3, activation='relu'))
    model.add(layers.MaxPooling2D(2))
    model.add(layers.Conv2D(16, 3, activation='relu'))
    model.add(layers.Flatten())
    model.add(layers.Dense(10))
    self.model = model

  def call(self, x):
    return self.model(x)

def toy_dataset(inputs, labels):
  return tf.data.Dataset.from_tensor_slices(
    dict(x=inputs, y=labels)).repeat().batch(32)

def train_step(net, example, optimizer):
  """Trains `net` on `example` using `optimizer`."""
  with tf.GradientTape() as tape:
    output = net(example['x'], training=True)
    loss = tf.losses.SparseCategoricalCrossentropy(from_logits=True)(example['y'], output)
  variables = net.trainable_variables
  gradients = tape.gradient(loss, variables)
  optimizer.apply_gradients(zip(gradients, variables))
  return loss

def getoptimizer():
  learning_rate = tf.optimizers.schedules.ExponentialDecay(
              tf.Variable(0.001), decay_steps=10000, decay_rate=0.5, staircase=False)
  """ save other parameters as well, from (https://gist.github.com/yoshihikoueno/4ff0694339f88d579bb3d9b07e609122)"""
  optimizer = tf.keras.optimizers.Adam(
                        learning_rate=learning_rate,
                        beta_1=tf.Variable(0.9),
                        beta_2=tf.Variable(0.999),
                        epsilon=tf.Variable(1e-7), amsgrad=True)
  
  optimizer = tfa.optimizers.MovingAverage(optimizer, average_decay=0.99)
  optimizer.iterations
  optimizer.decay = tf.Variable(0.0)
  return optimizer

def setup(log_dir, save_sub_optimizer=False):
  optimizer = getoptimizer()
  dataset = toy_dataset(train_images, train_labels)
  iterator = iter(dataset)

  model = Model()

  if save_sub_optimizer: 
    ckpt = tf.train.Checkpoint(step=tf.Variable(1), optimizer=optimizer, _optimizer = optimizer._optimizer,  model=model)
  else:
    ckpt = tf.train.Checkpoint(step=tf.Variable(1), optimizer=optimizer, model=model)

  manager = tf.train.CheckpointManager(ckpt, log_dir, max_to_keep=2)

  status = ckpt.restore(manager.latest_checkpoint) # Restore latest checkpoint
  if manager.latest_checkpoint:
    print("Restored from {}".format(manager.latest_checkpoint))
  else:
    print("Initializing from scratch.")

  return model, iterator, ckpt, manager, optimizer

def train_and_checkpoint(model, iterator, ckpt, manager, opt):

  for i in range(0, 500):
    example = next(iterator)
    loss = train_step(net, example, opt)
    ckpt.step.assign_add(1)
    if int(ckpt.step) % 50 == 0:
      save_path = manager.save()
      print("Saved checkpoint for step {}: {}".format(int(ckpt.step), save_path))
      print("loss {:1.2f}".format(loss.numpy()))

"""
## WITHOUT explicitly saving the Adam Optimzer inside MovingAverage from tf.addons
"""

log_dir = "/content/log/test"
net, iterator, ckpt, manager, opt = setup(log_dir, save_sub_optimizer=False)
train_and_checkpoint(net, iterator, ckpt, manager, opt)

del net
del iterator
del ckpt
del manager
del opt

net, iterator, ckpt, manager, opt = setup(log_dir, save_sub_optimizer=False)
net(np.zeros((1,28,28,1))) # fake input so variables get created
grad_vars = net.trainable_weights
zero_grads = [tf.zeros_like(w) for w in grad_vars]
opt.apply_gradients(zip(zero_grads, grad_vars)) # Apply zero gradients which don't do nothing with Adam (not entirely correct but does the trick for this purpose)

print(opt._optimizer.variables()[-1])

reader = tf.train.load_checkpoint(log_dir)
shape_from_key = reader.get_variable_to_shape_map()

print(sorted(shape_from_key.keys()))

del net
del iterator
del ckpt
del manager
del opt

"""
## WITH explicitly saving the Adam Optimzer inside MovingAverage from tf.addons

"""

log_dir = "/content/log/test2"
move_mean = True
net, iterator, ckpt, manager, opt = setup(log_dir, save_sub_optimizer=True)
train_and_checkpoint(net, iterator, ckpt, manager, opt)

del net
del iterator
del ckpt
del manager
del opt

net, iterator, ckpt, manager, opt = setup(log_dir, save_sub_optimizer=True)
net(np.zeros((1,28,28,1))) # fake input
grad_vars = net.trainable_weights
zero_grads = [tf.zeros_like(w) for w in grad_vars]
opt.apply_gradients(zip(zero_grads, grad_vars))

print(opt._optimizer.variables()[-1])

reader = tf.train.load_checkpoint(log_dir)
shape_from_key = reader.get_variable_to_shape_map()

print(sorted(shape_from_key.keys()))

`

**Other info / logs**

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by bhack** [comment]
Can you try with the last addons version?

**Comment by gowithdaflo** [comment]
I tested it with tfa version 0.12.0 and it works.

**Comment by gowithdaflo** [comment]
I also tested version 0.11.2 in combination with tf 2.2 and the same problem occurs. But seems to be fixed for tfa version >=0.12
